File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.cpp
class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        int n = s.size();
        int l = 0, count1 = 0;
        string ans = "";

        for (int r = 0; r < n; r++) {
            if (s[r] == '1') count1++;

            while (count1 == k) {
                string sub = s.substr(l, r - l + 1);
                if (ans.empty() || sub.size() < ans.size() || (sub.size() == ans.size() && sub < ans)) {
                    ans = sub;
                }
                if (s[l] == '1') count1--;
                l++;
            }
        }
        return ans;
    }
};

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution:
    def shortestBeautifulSubstring(self, s: str, k: int) -> str:

        #edge case
        if(s.count("1")<k):
            return ""
        
        ans=""
        minlen=float('inf')
        l,ones=0,0

        for r in range(len(s)):
            if(s[r]=='1'):
                ones+=1
            while(ones==k):
                clen=r-l+1
                csub=s[l:r+1]
                if(clen<minlen):
                    minlen=clen
                    ans=s[l:r+1]
                elif(clen==minlen and csub<ans):
                    ans=csub
                
                if(s[l]=='1'):
                    ones-=1
                l+=1
        return ans



        
        